### ISSUE TITLE
viostor win10 64-bit pro higher CPU/time in reading

### ISSUE THREAD

**Issue by absinthetized** [YanVugenfirer/kvm-guest-drivers-windows#62]
hello,

we are testing a win10 VM with virtio **0.1.113**. While all runs perfectly, it seems that we have performance regressions wrt win 8.1 64-bit pro. This is an early report, and it is rude, but it has prompted out by a simple "dir" command.

Additional info are provided at given links at end of report.

This is what I can recall to be relavant: 

- host is an hp dl160gen9 with centos 7.x, virtualization is given by **qemu-kvm 1.5.3-105.el7_2.3** (centos virt sig package). kvm-intel is the module.
- storage is **local** and is provided by soft raid5.
- both machines (win8.1 and win10) are fresh installs and have a 80GB virtio disk. **raw** storage on top of a **thick LVM** (same volume group). We have tried various storage strategies (from read cache to read/write cache with threads) and we have finally landed again to defaults.
- both machines have **2 vcpu**. topology is 2 vcpu with 1 vcores each. CPU is the same of host.
(config will show more cpu but desktop win versions limit available cpus to 2 sockets).

What we see is an higher time and CPU usage when accessing disk in read (write actually not tested yet). In order to have an idea of what happens, we have fired up a power-shell on both machines. The issued cmd is:

`cd c:\windows\system32`
`Measure-Command {dir}`

both win 8.1 and win 10 have really comparable system32 sizes but win 10 requires 350ms vs 230 of win8.1. test has been repeated (by hand) 20 times given consistent results.
At the same time the total cpu usage remains around 20% on win 8.1, while it jump up to 30% on win 10.

Is this to be expected at the given devel stage, or can we provide additional info on the topic (any suggestion is welcome)?

bye,
M

[win 8.1 full config](http://pastebin.com/kAfnDXRL)
[win 10 full config](http://pastebin.com/kd7H98FK)

**Comment by vrozenfe** [comment]
Can you please give a try to build 117?
Thanks,
Vadim.

**Comment by absinthetized** [comment]
hello. just tested. same behavior.

**Comment by ladipro** [comment]
Using 'dir' as an I/O benchmark is far from ideal. The command may be slower on Win10 for many reasons, especially since you're using it on a directory which, though similar in size, has a different content.

Could you please try a dedicated I/O benchmarking program, for example https://github.com/axboe/fio ?

**Comment by absinthetized** [comment]
tried a preliminary test with totally "random" settings:
```
; -- start job file --`
[read]
rw=read
filesize=474k
;size=1460320k
numjobs=1
nrfiles=3084
; -- end job file --
```

num of files and final overall size are just those obtained looking at the infamous windows/system32 dir output (on win8.1). I don't even know if serial read is the corrent approach to start.
here are results:
win8.1:
```
fio: this platform does not support process shared mutexes, forcing use of threads. Use the 'thread' option to get rid of this warning.
read: (g=0): rw=read, bs=4K-4K/4K-4K/4K-4K, ioengine=windowsaio, iodepth=1
fio-2.6
Starting 1 thread

read: (groupid=0, jobs=1): err= 0: pid=2904: Wed Apr 27 12:26:18 2016  read : io=1421.6MB, bw=38887KB/s, iops=9721, runt= 37433msec
    slat (usec): min=20, max=28847, avg=36.33, stdev=128.54
    clat (usec): min=6, max=12848, avg=45.51, stdev=51.77
     lat (usec): min=30, max=28891, avg=81.84, stdev=138.92
    clat percentiles (usec):
     |  1.00th=[   27],  5.00th=[   29], 10.00th=[   30], 20.00th=[   37],
     | 30.00th=[   38], 40.00th=[   39], 50.00th=[   41], 60.00th=[   48],
     | 70.00th=[   51], 80.00th=[   53], 90.00th=[   55], 95.00th=[   60],
     | 99.00th=[   91], 99.50th=[  118], 99.90th=[  247], 99.95th=[  274],
     | 99.99th=[ 1256]
    bw (KB  /s): min=22898, max=46216, per=100.00%, avg=38922.51, stdev=3184.54
    lat (usec) : 10=0.23%, 20=0.25%, 50=63.23%, 100=35.45%, 250=0.76%
    lat (usec) : 500=0.06%, 750=0.01%, 1000=0.01%
    lat (msec) : 2=0.01%, 4=0.01%, 10=0.01%, 20=0.01%
  cpu          : usr=5.34%, sys=66.79%, ctx=0, majf=0, minf=0
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued    : total=r=363912/w=0/d=0, short=r=0/w=0/d=0, drop=r=0/w=0/d=0
     latency   : target=0, window=0, percentile=100.00%, depth=1

Run status group 0 (all jobs):
   READ: io=1421.6MB, aggrb=38886KB/s, minb=38886KB/s, maxb=38886KB/s, mint=37433msec, maxt=37433msec
```
win 10:
```
fio: this platform does not support process shared mutexes, forcing use of threads. Use the 'thread' option to get rid of this warning.
read: (g=0): rw=read, bs=4K-4K/4K-4K/4K-4K, ioengine=windowsaio, iodepth=1
fio-2.6
Starting 1 thread

read: (groupid=0, jobs=1): err= 0: pid=6032: Wed Apr 27 12:27:41 2016  read : io=1421.6MB, bw=46407KB/s, iops=11601, runt= 31367msec
    slat (usec): min=21, max=5718, avg=29.19, stdev=23.07
    clat (usec): min=6, max=8994, avg=37.23, stdev=24.86
     lat (usec): min=28, max=9023, avg=66.42, stdev=34.27
    clat percentiles (usec):
     |  1.00th=[   28],  5.00th=[   34], 10.00th=[   34], 20.00th=[   35],
     | 30.00th=[   35], 40.00th=[   35], 50.00th=[   35], 60.00th=[   36],
     | 70.00th=[   36], 80.00th=[   36], 90.00th=[   40], 95.00th=[   43],
     | 99.00th=[   76], 99.50th=[  116], 99.90th=[  231], 99.95th=[  251],
     | 99.99th=[  422]
    bw (KB  /s): min=20240, max=48224, per=100.00%, avg=46497.03, stdev=3543.95
    lat (usec) : 10=0.36%, 20=0.12%, 50=96.19%, 100=2.52%, 250=0.75%
    lat (usec) : 500=0.04%, 750=0.01%, 1000=0.01%
    lat (msec) : 2=0.01%, 4=0.01%, 10=0.01%
  cpu          : usr=0.00%, sys=70.14%, ctx=0, majf=0, minf=0
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued    : total=r=363912/w=0/d=0, short=r=0/w=0/d=0, drop=r=0/w=0/d=0
     latency   : target=0, window=0, percentile=100.00%, depth=1

Run status group 0 (all jobs):
   READ: io=1421.6MB, aggrb=46406KB/s, minb=46406KB/s, maxb=46406KB/s, mint=31367msec, maxt=31367msec

```
I think most relevant items here are:
- win 8.1: aggrb=38886KB/s with cpu: usr=5.34%, sys=66.79%
- win 10:  aggrb=46406KB/s with cpu: usr=0.00%, sys=70.14%

TBH, if this can be a proper benchmark I don't see anything wrong and our perception was probably induced by something **not** related to virtio storage.

Let me know If do you need anything else to be tested (as I'm going have a bit of time to compile a number of fio jobs, just use this chance).

**Comment by YanVugenfirer** [comment]
There are several performance tools that can be used to understand what is causing CPU consumption. Starting from Intel's vTune (there is 30days trial version) to MS xperf tool that is part of the SDK.

**Comment by vrozenfe** [comment]
I wouldn't recommend running xperf on non-development setup :)
PerfMon or Process Explorer could be enough for monitoring high CPU 
usage thread(s).

Matteo, could you try activating hyper-v enlightenment features (especially hv_time)
to see if it makes any difference?

Best regards,
Vadim.

**Comment by absinthetized** [comment]
ok,

hyperv emulation enabled by means described in [this post](http://blog.wikichoon.com/2014/07/enabling-hyper-v-enlightenments-with-kvm.html).

as first I've enabled only the features (those not mentioned explicitly by @vrozenfe ):
neither fio nor "powershell" seemed to have perceived the change. 

of course I've checked the xml config for the VM and hyperv stuff was in.

**then** I've also enabled the clock (and checked for it being in the xml):
fio now jumps up to 60.000/65.000 KB/s (tested twice) and CPU is dropped down around 40%, far more performing and far less stressing than "plain" win 8.1 without enlightenment.

about the rude power-shell 'dir': now we have comparable cpu and timings. so I **guess** hyperv clock totally solved the issue.

while I can understand the bias in testing virt storage with a simple dir, my question is: what kind of overhead is there to let a file listing command perform worse than a io testing suite? I mean: before hyperv clock, fio returned comparable results on both machines, but 'dir' command resource usage dropped on win10 only after hyperv clock was there. 

[strange to me, but I'm just an under-average python developer, really ignoring FS and kernel internals :)]

thank you!
M

PS: I've not closed the ticket just in case you want to comment/add anything I can do.
